src/QueryEditor.tsx

```diff
@@ -1,47 +1,74 @@
-import { Column, DataframeQuery, QueryColumn } from './types';
-import { Button, InlineField, Input, MultiSelect } from '@grafana/ui';
+import { Column, DataframeQuery, isValidQuery, QueryColumn } from './types';
+import { InlineField, InlineSwitch, Input, MultiSelect, Select } from '@grafana/ui';
+import { decimationMethods, defaultDecimationMethod } from './constants';
 
 type Props = QueryEditorProps<DataFrameDataSource, DataframeQuery>;
 
-export const QueryEditor: React.FC<Props> = ({ app, query, datasource, onChange, onRunQuery }) => {
+export const QueryEditor: React.FC<Props> = ({ query, datasource, onChange, onRunQuery }) => {
   const tableMetadata = useAsync(async () => {
     return query.tableId ? await datasource.getTableMetadata(query.tableId) : null;
   }, [query.tableId]);
 
+  const runQueryIfValid = () => isValidQuery(query) && onRunQuery();
+
+  const handleQueryChange = (value: DataframeQuery, runQuery?: boolean) => {
+    onChange(value);
+    if (runQuery) {
+      runQueryIfValid();
+    }
+  };
+
   const handleIdInputBlur = (e: React.FocusEvent<HTMLInputElement>) => {
     if (query.tableId !== e.currentTarget.value) {
-      onChange({ ...query, tableId: e.currentTarget.value, columns: [] });
+      handleQueryChange({ ...query, tableId: e.currentTarget.value, columns: [] });
     }
   };
 
   const handleColumnChange = (items: Array<SelectableValue<string>>) => {
     const columns = items.map(({ value, dataType, columnType }) => ({ name: value!, dataType, columnType }));
-    onChange({ ...query, columns });
+    handleQueryChange({ ...query, columns });
   };
 
   return (
     <>
       <InlineField label="Id" error="Table does not exist" invalid={!!tableMetadata.error}>
         <Input defaultValue={query.tableId} width={26} onBlur={handleIdInputBlur} />
       </InlineField>
-      <InlineField label="Columns">
+      <InlineField label="Columns" tooltip="The columns to include in the response data.">
         <MultiSelect
           isLoading={tableMetadata.loading}
           options={columnsToOptions(tableMetadata.value?.columns)}
           onChange={handleColumnChange}
+          onBlur={runQueryIfValid}
           value={columnsToOptions(query.columns)}
         />
       </InlineField>
-      {app !== CoreApp.Explore && (
-        <Button size="sm" icon="play" className={css({ marginTop: 20 })} onClick={onRunQuery}>
-          Run query
-        </Button>
-      )}
+      <InlineField label="Decimation" tooltip="Specifies the method used to decimate the data.">
+        <Select
+          options={decimationMethods}
+          onChange={(item) => handleQueryChange({ ...query, decimationMethod: item.value }, true)}
+          value={query.decimationMethod ?? defaultDecimationMethod}
+        />
+      </InlineField>
+      <InlineField label="Filter nulls" tooltip="Filter out null and NaN values before decimating the data.">
+        <InlineSwitch
+          value={query.filterNulls}
```

Comment:
If this doesn't already, should this default to true?

Reply:
No, this is intentionally off by default. This is so the first query a user makes in the Explore view gets the whole range of the data. Talking to Mark, the long term goal is to have this option turn itself on if the user interacts with the graph, but I couldn't find a straightforward way to do this. We'll have to see if we can improve it later.